libuser/src/lib.rs

```diff
@@ -4,18 +4,53 @@
 
 #![no_std]
 
-use core::arch::{asm, global_asm};
+mod hypcalls;
+
+pub use crate::hypcalls::*;
+use core::arch::global_asm;
 
 global_asm!(include_str!("start.S"));
 
-// Loop making ecalls as the kernel will kill the task on an ecall (the only syscall supported is
-// `exit`).
 #[panic_handler]
-fn panic(_info: &core::panic::PanicInfo) -> ! {
-    // Safe to make an ecall that won't return.
-    unsafe {
-        loop {
-            asm!("ecall");
+fn panic(info: &core::panic::PanicInfo) -> ! {
+    println!("panic : {:?}", info);
+    hyp_panic();
+}
+
+/// Writer for printing characters through the hypervisor via the putchar hypcall.
+pub struct UserWriter {}
+
+impl core::fmt::Write for UserWriter {
+    fn write_str(&mut self, s: &str) -> core::fmt::Result {
+        for c in s.chars() {
+            // Ignore errors for putchar.
+            let _ = hyp_putchar(c);
         }
+        Ok(())
     }
 }
+
+/// Global reference for `print` implementation.
+pub static mut UMODEWRITER: UserWriter = UserWriter {};
```

Comment:
nit: `UMODE_WRITER` or `USER_WRITER`, etc.

Reply:
Ack.

Reply:
Fixed. This was actually a spurious line, I didn't need this static variable anymore -- rust missed that being declared public in a library.